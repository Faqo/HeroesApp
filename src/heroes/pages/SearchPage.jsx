import { useLocation, useNavigate } from "react-router-dom";
import { useForm } from "../../hooks/useForm"
import { HeroCard } from "../components/HeroCard"
import queryString from 'query-string';
import { getHeroesByName } from "../helpers";

export const SearchPage = () => {

  const navigate = useNavigate();
  const location = useLocation();

  const { q=''} = queryString.parse(location.search);

  const heroes = getHeroesByName(q);

  const {searchText, onInputChange} = useForm({
    searchText: q,
  });

  const onSearchSubmit = (e) =>{
    e.preventDefault();
    if(searchText.trim().length <=1) return;
    // console.log({searchText});
    navigate(`?q=${searchText}`);
  };

  return (
    <>
      <h1>Search</h1>
      <hr />
      <div className="row">
        <div className="col-5">
          <h4>Searching</h4>
          <hr />
          <form onSubmit={onSearchSubmit}>
            <input 
              type="text"
              placeholder="Search a hero"
              className="form-control"
              name="searchText"
              autoComplete="off"
              value={searchText}
              onChange={onInputChange}
              />
              <button 
                className="btn btn-outline-primary mt1">
                  search
              </button>
          </form>
        </div>
        <div className="col-7">

          <h4>Results</h4>
          <hr />

          <div className="alert alert-primary">
            Search a hero
          </div>
          
          <div className="alert alert-danger">
            no hero with <b>{ q }</b>
          </div>
          {
            heroes.map(hero => (
              <HeroCard key={hero.id} {...hero}/> 
            ))
          }

        </div>
      </div>
    </>
  )
}