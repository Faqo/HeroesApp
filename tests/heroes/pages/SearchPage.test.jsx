import { render, screen, fireEvent } from "@testing-library/react";
import { MemoryRouter } from 'react-router-dom';
import { SearchPage } from "../../../src/heroes/pages/SearchPage";


const mockUseNavigate = jest.fn()

jest.mock('react-router-dom', ()=>({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockUseNavigate
})); 

describe('Pruebas enSearchPage', () => {

    beforeEach(()=> jest.clearAllMocks() );

    test('Debe de mostrarse correctamente con valores por defecto ', () => {
        const { conteiner } = render(
                                    <MemoryRouter>
                                        <SearchPage />
                                    </MemoryRouter>
                                );
        expect(conteiner).toMatchSnapshot();
    });
    test('Debe de mostrar a batman y el input con el querystring ', () => {
        render(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <SearchPage />
            </MemoryRouter>
        );
        const input = screen.getByRole('textbox');
        expect(input.value).toBe('batman');

        const img = screen.getByRole('img');
        expect(img.src).toContain('/assets/heroes/dc-batman.jpg');
    
        const alert = screen.getByLabelText('search-label');
        expect(alert.style.display).toBe('none');

        const alertDanger = screen.getByLabelText('error-label');
        expect(alertDanger.style.display).toBe('none');
    });
    test('debe de mostrar un error si no encuentra el heroe(batman2)', () => {
        render(
            <MemoryRouter initialEntries={['/search?q=batman2']}>
                <SearchPage />
            </MemoryRouter>
        );
        const alertDanger = screen.getByLabelText('error-label');
        expect(alertDanger.style.display).toBe('');
        expect(alertDanger.textContent).toContain('batman2');
    });

    test('debe de llamar el navigate a la pantalla nueva', () => {
        render(
            <MemoryRouter initialEntries={['/search']}>
                <SearchPage />
            </MemoryRouter>
        );
        const inputValue = 'superman';
        const input = screen.getByRole('textbox');
        // const form = screen.getByRole('form');

        fireEvent.input(input, {target: {name: "searchText",value: inputValue}});
        fireEvent.submit( input ); 
        // fireEvent.submit( form ); 

        expect(mockUseNavigate).toHaveBeenCalledWith(`?q=${inputValue}`)

    });
});