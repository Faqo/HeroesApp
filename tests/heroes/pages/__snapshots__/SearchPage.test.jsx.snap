// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas enSearchPage Debe de mostrarse correctamente con valores por defecto  1`] = `undefined`;
